c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -205,14 +205,14 @@ impl<'tcx> TypeChecker<'tcx, '_> {
     /// type has a pointer, this function unifies the `PointerId`s that `lty1` and `lty2` have at
     /// that position.  For example, given `lty1 = *mut /*l1*/ *const /*l2*/ u8` and `lty2 = *mut
     /// /*l3*/ *const /*l4*/ u8`, this function will unify `l1` with `l3` and `l2` with `l4`.
-    fn do_unify(&mut self, lty1: LTy<'tcx>, lty2: LTy<'tcx>) {
-        let ty1 = lty1.ty;
-        let ty2 = lty2.ty;
-        assert!(are_transmutable(
-            self.acx.tcx().erase_regions(ty1),
-            self.acx.tcx().erase_regions(ty2),
-        ), "types not transmutable (compatible), so PointerId unification cannot be done: {ty1:?} !~ {ty2:?}");
-        for (sub_lty1, sub_lty2) in lty1.iter().zip(lty2.iter()) {
+    fn do_unify(&mut self, pl_lty: LTy<'tcx>, rv_lty: LTy<'tcx>) {
+        let pl_ty = pl_lty.ty;
+        let rv_ty = rv_lty.ty;
+        assert!(is_transmutable_to(
+            self.acx.tcx().erase_regions(rv_ty),
+            self.acx.tcx().erase_regions(pl_ty),
+        ), "types not transmutable (compatible), so PointerId unification cannot be done: *{rv_ty:?} as *{pl_ty:?}");
+        for (sub_lty1, sub_lty2) in pl_lty.iter().zip(rv_lty.iter()) {
```

Comment:
I'm confused here: are `pl_lty` née `lty1` and `rv_lty` née `lty2` the (possibly) pointer types, or types that have already been dereferenced? If the former, I don't see how their transmute compatibility can be handled by `is_transmutable_to` née `are_transmutable`, because that function doesn't do any dereferencing. If they've already been dereferenced, then the comment giving the example of `lty1 = *mut /*l1*/ *const /*l2*/ u8` seems misleading.

In any event, update the `lty1`/`lty2` names in the comment.

Reply:
> I'm confused here: are `pl_lty` née `lty1` and `rv_lty` née `lty2` the (possibly) pointer types, or types that have already been dereferenced? If the former, I don't see how their transmute compatibility can be handled by `is_transmutable_to` née `are_transmutable`, because that function doesn't do any dereferencing. If they've already been dereferenced, then the comment giving the example of `lty1 = *mut /*l1*/ *const /*l2*/ u8` seems misleading.

`pl_ty` and `rv_ty` are the possibly pointer types, but `tcx.erase_regions` strips the pointer (if it's there) from what I can tell.

> In any event, update the `lty1`/`lty2` names in the comment.

Thanks for catching that, I'll do that.

Reply:
I could be wrong, but I don't think `erase_regions` strips an outer pointer: looking at [this source](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_middle/ty/erase_regions.rs.html), I would expect `fold_ty` to do dispatch on the outer kind, but it only either dispatches to itself via the generic visitor (`super_fold_with`), or to a new instance of itself (`self.tcx.erase_regions_ty(ty)`).

Reply:
> In any event, update the `lty1`/`lty2` names in the comment.

Fixed in f05df654f7d3a12d0d104419b03fefd24639fbc9.

Reply:
> I could be wrong, but I don't think `erase_regions` strips an outer pointer

Yeah, you're right.  I misunderstood.

It seems that the `pl_ty` and `rv_ty` types passed to `do_unify` have already been dereferenced, which happens in `do_equivalence_nested` and other call sites.  I'm not sure what the example means.  @spernsteiner?

Reply:
`*const *mut *mut u8` is a legal type.  `do_equivalence_nested` will strip off the outermost `*const` and pass `*mut *mut u8` to `do_unify`.